src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <exception>
+#include <map>
+#include <set>
+#include <pthread.h>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cudf_jni_apis.hpp>
+
+namespace {
+
+constexpr char const *RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
+
+class thread_state {
+public:
+    bool retry_oom_injected = false;
+    bool split_and_retry_oom_injected = false;
+    long task_id = -1;
+};
+
+class rollback {
+    using T = const std::function<void()>;
+public:
+    rollback(T & on_error): on_error(on_error) {}
+
+    ~rollback() {
+      if (std::uncaught_exceptions() > 0) {
+        on_error();
+      }
+    }
+private:
+    T on_error;
+};
+
+class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+public:
+  spark_resource_adaptor(JNIEnv *env, rmm::mr::device_memory_resource *mr)
+      : resource{mr} {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw std::runtime_error("GetJavaVM failed");
+    }
+  }
+
+  rmm::mr::device_memory_resource *get_wrapped_resource() { return resource; }
+
+  bool supports_get_mem_info() const noexcept override { return resource->supports_get_mem_info(); }
+
+  bool supports_streams() const noexcept override { return resource->supports_streams(); }
+
+  void associate_thread_with_task(long thread_id, long task_id) {
+    std::scoped_lock lock(state_mutex);
+    if (threads.find(thread_id) != threads.end()) {
+      throw std::invalid_argument("a thread can only be added if it is in the unknown state");
+    }
+    threads.insert({thread_id, {false, false, task_id}});
+    {
+      rollback rb([this, thread_id]() {
```

Comment:
so when do we expect the rollback to trigger? e.g. why do we need to have this RAII object?

Perhaps a class header on `rollback` may help me.

Reply:
It is not about when do I expect this to happen. It is about the fact that an exception can happen. Any of the `insert` calls that happen after this could throw an exception. If that happens `threads` would point to something that is not inside of `task_to_threads`. The spec for `erase` says 

> Unless the container's [comparison object](https://cplusplus.com/map::key_comp) throws, this function never throws exceptions (no-throw guarantee).


So this code should be able to handle anything that is thrown at it. Do I think that this is going to happen in practice? No insert should only throw if memory allocation failed, which on modern Linux running a 64-bit process should really never happen thanks to virtual memory.

Reply:
This is for my own edification, I am just curious if the below would be equivalent. If so that's fine, I am just trying to understand if we prefer RAII for this, instead of try-catch-rethrow.

```
threads.insert({thread_id, {false, false, task_id}});
try {
  auto tasks_it = task_to_threads.find(task_id);
  if (tasks_it == task_to_threads.end()) {
    task_to_threads.insert({task_id, {thread_id}});
  } else {
    tasks_it->second.insert(thread_id);
  }
} catch (const std::exception& x) {
  threads.erase(thread_id);
  throw;
}
```

Reply:
I kind of prefer the try/catch block here.  The rollback is emulating finally, but it is is explicitly checking for uncaught exceptions, so isn't it equivalent to a catch?